File: src/cpp/Mirror.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
public:
    void Mirror(TreeNode *pRoot) {
        // 如果为空子树，直接返回
        if (pRoot == NULL) {
            return;
        }

        // 左右节点交换
        TreeNode *temp = NULL;
        temp = pRoot->left;
        pRoot->left = pRoot->right;
        pRoot->right = temp;

        // 左右子树递归
        Mirror(pRoot->left);
        Mirror(pRoot->right);
    }
};

File: src/java/Mirror.java
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    // 递归解法
    public void Mirror(TreeNode root) {
        if (root == null) return;

        if (root.left == null && root.right == null) return;

        TreeNode temp = root.left;
        root.left = root.right;
        root.right = temp;

        Mirror(root.left);
        Mirror(root.right);
    }
    // 非递归解法
    public void Mirror2(TreeNode root) {
        if (root == null) return;

        if (root.left == null && root.right == null) return;

        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.empty()) {
            TreeNode top = stack.pop();

            // swap
            if (top.left != null || top.right != null) {
                TreeNode temp = top.left;
                top.left = top.right;
                top.right = temp;
            }

            if (top.left != null) {
                stack.push(top.left);
            }

            if (top.right != null) {
                stack.push(top.right);
            }
        }
    }
}
